### PR TITLE
ARMOCP-411: Add ARM nodepool to Azure x86 Hosted Cluster

### PR DESCRIPTION
**Pull request by bryan-cox**
**What this PR does / why we need it**:
- Allows a user to add an ARM nodepool to an Azure x86 hosted cluster
- Updates Azure VM image creation to use image gallery

**Which issue(s) this PR fixes**
Fixes - [ARMOCP-411 - Add arm nodepool to x86 managed cluster (Azure)](https://issues.redhat.com/browse/ARMOCP-411)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

File: cmd/cluster/azure/create.go [hunk 1]
```diff
@@ -83,12 +84,27 @@ func applyPlatformSpecificsValues(ctx context.Context, exampleOptions *apifixtur
...
 		if err != nil {
 			return fmt.Errorf("failed to create infra: %w", err)
 		}
 	}
 
+	var instanceType string
+	if opts.AzurePlatform.InstanceType != "" {
+		instanceType = opts.AzurePlatform.InstanceType
+	} else {
+		// Aligning with Azure IPI instance type defaults
```
Comment (enxebre) [inline cmd/cluster/azure/create.go:99]: is there a link to this ipi defaults?

File: hypershift-operator/controllers/nodepool/azure.go [hunk 2]
```diff
@@ -55,11 +62,23 @@ func generateSSHPubkey() (string, error) {
...
+	if len(splitResourceGroupName) <= 0 {
+		return "", fmt.Errorf("failed to parse resource group name unique ID to retrieve boot image: %s", hcluster.Spec.Platform.Azure.ResourceGroupName)
+	} else {
+		bootImageGalleyID = splitResourceGroupName[len(splitResourceGroupName)-1]
+	}
+
+	switch nodepool.Spec.Arch {
+	case "amd64":
+		return fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/galleries/BootImageGallery_%s/images/RHCOS_Amd64/versions/1.0.0", hcluster.Spec.Platform.Azure.SubscriptionID, hcluster.Spec.Platform.Azure.ResourceGroupName, bootImageGalleyID), nil
+	case "arm64":
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/azure.go:77]: please use constants so we only write the string once

File: hypershift-operator/controllers/nodepool/azure.go [hunk 3]
```diff
@@ -55,11 +62,23 @@ func generateSSHPubkey() (string, error) {
...
+	splitResourceGroupName := strings.Split(hcluster.Spec.Platform.Azure.ResourceGroupName, "-")
+	if len(splitResourceGroupName) <= 0 {
+		return "", fmt.Errorf("failed to parse resource group name unique ID to retrieve boot image: %s", hcluster.Spec.Platform.Azure.ResourceGroupName)
+	} else {
+		bootImageGalleyID = splitResourceGroupName[len(splitResourceGroupName)-1]
+	}
+
+	switch nodepool.Spec.Arch {
+	case "amd64":
+		return fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/galleries/BootImageGallery_%s/images/RHCOS_Amd64/versions/1.0.0", hcluster.Spec.Platform.Azure.SubscriptionID, hcluster.Spec.Platform.Azure.ResourceGroupName, bootImageGalleyID), nil
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/azure.go:76]: this can be a parametarised function so you don't need to handcraft it again in the tests

File: api/fixtures/example.go [hunk 4]
```diff
@@ -71,6 +71,7 @@ type ExampleOptions struct {
 	AutoRepair                       bool
```
Comment (enxebre) [inline api/fixtures/example.go:71]: Please update commit messages to follow imperative mood https://github.com/openshift/hypershift/blob/main/docs/content/contribute/index.md e.g Add arm64 support for Azure NodePool

File: cmd/cluster/azure/create.go (hunk 1 above)
Comment (bryan-cox) [reply cmd/cluster/azure/create.go:99]: I had the wrong x86 one in there - should be Standard_D4s_v3 instead of v4. I'll fix that. For x86, the documentation is here - https://access.redhat.com/documentation/en-us/openshift_container_platform/4.11/html/installing/installing-on-azure. I don't see anything for arm in there but I selected the equivalent arm instance type based on vCPU, Memory, etc based on the specs for Standard_D4s_v3.

File: hypershift-operator/controllers/nodepool/azure.go (hunk 2 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/azure.go:77]: Updated this to use the same constants created in https://github.com/openshift/hypershift/pull/1594

File: api/fixtures/example.go (hunk 4 above)
Comment (bryan-cox) [reply api/fixtures/example.go:71]: This has been updated.

File: hypershift-operator/controllers/nodepool/azure.go (hunk 3 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/azure.go]: Yeah that is a good idea. The function has been updated.

**Comment by bryan-cox** [comment]
@enxebre - I just pushed some changes to cleanup/simplify the Azure infrastructure Run function. 

However, I wanted to ask about another change I was recently thinking of making. Creating a hosted cp on Azure seems to be slower than creating one on AWS. This is especially true when there is more than one RHCOS architecture type to upload and create VM image definitions for. Would it be desirable to have some concurrency put in for that? 

Basically, I was thinking of concurrently uploading the RHCOS and creating the VM image definition things for ARM and x86 at the same time with a `WaitGroup` rather than the current implementation of uploading/creating the VM image definition for one arch then sequentially working on the next arch.

**Comment by bryan-cox** [comment]
@enxebre - [Here](https://github.com/openshift/hypershift/pull/1755/commits/98e38a8f78fa760cd4e12228b0814d194339de6a) is what I meant about creating the RHCOS image containers/definitions concurrently. It saves about 10 minutes when creating a cluster with a multi-arch release image. I left it has its own commit for now. If we want to keep the concurrent code, I can squash it in with the previous commit, i.e, [Add arm64 support for Azure NodePool](https://github.com/openshift/hypershift/pull/1755/commits/efeb10315360b43f6451a4994adbb683321f5709).
